internal/services/mongocluster/mongo_cluster_resource.go

```diff
@@ -0,0 +1,561 @@
+)
+
+type MongoClusterResource struct{}
+
+var _ sdk.ResourceWithUpdate = MongoClusterResource{}
+
+var _ sdk.ResourceWithCustomizeDiff = MongoClusterResource{}
+
+type MongoClusterResourceModel struct {
+	Name                       string            `tfschema:"name"`
+	ResourceGroupName          string            `tfschema:"resource_group_name"`
+	Location                   string            `tfschema:"location"`
+	AdministratorLogin         string            `tfschema:"administrator_login"`
+	AdministratorLoginPassword string            `tfschema:"administrator_login_password"`
+	CreateMode                 string            `tfschema:"create_mode"`
+	ShardCount                 int64             `tfschema:"shard_count"`
+	SourceLocation             string            `tfschema:"source_location"`
+	SourceServerId             string            `tfschema:"source_server_id"`
+	ComputeTier                string            `tfschema:"compute_tier"`
+	HighAvailabilityMode       string            `tfschema:"high_availability_mode"`
+	PublicNetworkAccessEnabled bool              `tfschema:"public_network_access_enabled"`
+	PreviewFeatures            []string          `tfschema:"preview_features"`
+	StorageSizeInGb            int64             `tfschema:"storage_size_in_gb"`
+	Tags                       map[string]string `tfschema:"tags"`
+	Version                    string            `tfschema:"version"`
+}
+
+func (r MongoClusterResource) ModelObject() interface{} {
+	return &MongoClusterResourceModel{}
+}
+
+func (r MongoClusterResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return mongoclusters.ValidateMongoClusterID
+}
+
+func (r MongoClusterResource) ResourceType() string {
+	return "azurerm_mongo_cluster"
+}
+
+func (r MongoClusterResource) Arguments() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"name": {
+			ForceNew: true,
+			Required: true,
+			Type:     schema.TypeString,
+			ValidateFunc: validation.All(
+				validation.StringLenBetween(3, 40),
+				validation.StringMatch(
+					regexp.MustCompile(`^[a-z0-9]+(-[a-z0-9]+)*$`),
+					"The name contains only lowercase letters, numbers and hyphens.",
+				),
+			),
+		},
+
+		"resource_group_name": commonschema.ResourceGroupName(),
+
+		"location": commonschema.Location(),
+
+		"administrator_login": {
+			Type:         schema.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+
+		"create_mode": {
+			Type:     schema.TypeString,
+			Optional: true,
+			ForceNew: true,
+			Default:  string(mongoclusters.CreateModeDefault),
+			// Confirmed with service team the 'Default' and `GeoReplica` are the only accepted value currently, other values will be supported later.
+			ValidateFunc: validation.StringInSlice([]string{
+				string(mongoclusters.CreateModeDefault),
+				string(mongoclusters.CreateModeGeoReplica),
+			}, false),
+		},
+
+		"preview_features": {
```

Comment:
Elements in list can be duplicated. What will happen if duplicated elements are sent to Azure?
Will Azure reorder the element sequence in response if more than 1 elements are sent to Azure?

Reply:
If duplicated elements are sent to Azure, the API returns a 400 Bad Request with an error message for duplicate array value.

Since there is currently only one possible value for `preview_features`(it could not be deleted once specified). I am not sure whether the API would reorder the elements.

Reply:
Could you add the validation for the duplicated values in the `customizeDiff` function so that such error can be found in plan stage?

Reply:
Fixed.